core-rust/state-manager/src/staging/cache.rs

```diff
@@ -147,20 +166,74 @@ impl ExecutionCache {
             }
         };
     }
+
+    fn process<S: RootStore>(
+        store: &StagedStore<S>,
+        transaction_receipt: TransactionReceipt,
+    ) -> ProcessedResult {
+        let hash_changes = match &transaction_receipt.result {
+            TransactionResult::Commit(commit) => commit
+                .state_updates
+                .up_substates
+                .iter()
+                .map(|(id, value)| {
+                    (
+                        id.clone(),
+                        Some(hash(scrypto_encode(&value.substate).unwrap())),
+                    )
+                })
+                .collect::<Vec<(SubstateId, Option<Hash>)>>(),
+            TransactionResult::Reject(_) | TransactionResult::Abort(_) => Vec::new(),
+        };
+        let mut collector = CollectingTreeStore::new(store);
+        let state_version = store.overlay.state_version;
+        let root_hash = put_at_next_version(&mut collector, state_version, &hash_changes);
+        ProcessedResult::new(root_hash, transaction_receipt, collector.diff)
+    }
+}
+
+struct CollectingTreeStore<'s, S: ReadableTreeStore> {
+    readable_delegate: &'s S,
+    diff: HashTreeDiff,
+}
+
+impl<'s, S: ReadableTreeStore> CollectingTreeStore<'s, S> {
+    pub fn new(readable_delegate: &'s S) -> Self {
+        Self {
+            readable_delegate,
+            diff: HashTreeDiff::new(),
+        }
+    }
 }
 
-pub struct StagedSubstateStore<'s, S: ReadableSubstateStore> {
+impl<'s, S: ReadableTreeStore> ReadableTreeStore for CollectingTreeStore<'s, S> {
+    fn get_node(&self, key: &NodeKey) -> Option<TreeNode> {
+        self.readable_delegate.get_node(key)
+    }
+}
+
+impl<'s, S: ReadableTreeStore> WriteableTreeStore for CollectingTreeStore<'s, S> {
+    fn insert_node(&mut self, key: &NodeKey, node: TreeNode) {
```

Comment:
Very minor, but I'd probably expect all stores to take ownership of the key on insert - so this should possibly be an owned key here in the interface (as that defers the cloning to the parent layer and might be more resistant to unnecessary cloning)